WordPress/Classes/ViewRelated/Blog/Site Picker/BlogList/BlogListViewModel.swift

```diff
@@ -0,0 +1,239 @@
+
+    private let contextManager: ContextManager
+    private let eventTracker: EventTracker
+
+    init(contextManager: ContextManager = ContextManager.sharedInstance(),
+         eventTracker: EventTracker = DefaultEventTracker()
+    ) {
+        self.contextManager = contextManager
+        self.eventTracker = eventTracker
+        super.init()
+        setupFetchedResultsControllers()
+    }
+
+    func updateSearchText(_ newText: String) {
+        if newText.isEmpty {
+            searchSites = allBlogs.compactMap(BlogListView.Site.init)
+        } else {
+            searchSites = allBlogs
+                .filter {
+                    $0.url?.lowercased().contains(newText.lowercased()) == true
+                    || $0.title?.lowercased().contains(newText.lowercased()) == true
+                }
+                .compactMap(BlogListView.Site.init)
+        }
+    }
+
+    func togglePinnedSite(siteID: NSNumber?) {
+        guard let siteID, let blog = allBlogs.first(where: { $0.dotComID == siteID }) else {
+            return
+        }
+        let isCurrentlyPinned = blog.pinnedDate != nil
+
+        trackPinned(blog: blog)
+        if isCurrentlyPinned {
+            moveRecentPinnedSiteToRemainingSitesIfNeeded(pinnedBlog: blog)
+        }
+
+        blog.pinnedDate = isCurrentlyPinned ? nil : Date()
+
+        contextManager.saveContextAndWait(contextManager.mainContext)
+    }
+
+    func siteSelected(siteID: NSNumber?) {
+        guard let siteID, let blog = allBlogs.first(where: { $0.dotComID == siteID }) else {
+            return
+        }
+
+        trackSiteSelected(blog: blog)
+
+        blog.lastUsed = Date()
+
+        updateExcessRecentBlogsIfNeeded(selectedSiteID: siteID)
+
+        contextManager.saveContextAndWait(contextManager.mainContext)
+    }
+
+    private func trackPinned(blog: Blog) {
+        eventTracker.track(
+            .siteSwitcherPinUpdated,
+            properties: [
+                "blog_id": blog.dotComID ?? "",
+                "pinned": blog.pinnedDate == nil
+            ]
+        )
+    }
+
+    private func trackSiteSelected(blog: Blog) {
+        let sectionName: String
+
+        if blog.pinnedDate != nil {
+            sectionName = "pinned"
+        } else if blog.lastUsed != nil {
+            sectionName = "recent"
+        } else {
+            sectionName = "all"
+        }
+
+        eventTracker.track(
+            .siteSwitcherSiteTapped,
+            properties: [
+                "section": sectionName,
+            ]
+        )
+    }
+
+    private static func filteredAllRemainingSites(allBlogs: [Blog]) -> [BlogListView.Site] {
+        allBlogs.filter({ $0.pinnedDate == nil && $0.lastUsed == nil }).compactMap(BlogListView.Site.init)
+    }
+
+    private func moveRecentPinnedSiteToRemainingSitesIfNeeded(pinnedBlog: Blog) {
+        if let recentBlogs = recentSitesController?.fetchedObjects,
+           recentBlogs.count == 8 {
+            pinnedBlog.lastUsed = nil
```

Comment:
This was to resolve the issue you posted before. Since I filter with keypath whether a blog has lastDate or not. If it has lastDate but doesn't fall into the first 8, it would still be filtered out. Also I thought it is better to treat them as they're not recent rather than finding a way to not filter out beyond 8 from the all sites list. It was going to require additional filtering which felt more error prone.

Reply:
Okay, I understand 👍

The downside to this is the issue I mentioned here: https://github.com/wordpress-mobile/WordPress-iOS/pull/22833#pullrequestreview-2017646833

Plus, it messes with the definition of the `lastUsed` property. This property can be used elsewhere in the future and this logic will make its behavior unexpected.